scripts/base/protocols/redis/main.zeek

```diff
@@ -0,0 +1,254 @@
+	# pairs.
+	if ( command?$known && command$known == KnownCommand_CLIENT )
+		{
+		# All 3 CLIENT commands we care about have 3 elements
+		if ( |command$raw| == 3 )
+			{
+			if ( to_lower(command$raw[2]) == "on" )
+				{
+				# If the last range is open, close it here. Otherwise, noop
+				if ( |c$redis_state$no_response_ranges| > 0 )
+					{
+					local range = c$redis_state$no_response_ranges[|c$redis_state$no_response_ranges|
+					    - 1];
+					if ( |range| == 1 )
+						{
+						range += c$redis_state$current_request;
+						}
+					}
+				}
+			if ( to_lower(command$raw[2]) == "off" )
+				{
+				# Only add a new interval if the last one is closed
+				if ( is_last_interval_closed(c) )
+					{
+					c$redis_state$no_response_ranges += vector(c$redis_state$current_request);
+					}
+				}
+			if ( to_lower(command$raw[2]) == "skip" )
+				{
+				if ( is_last_interval_closed(c) )
+					# It skips this one and the next one
+					c$redis_state$no_response_ranges += vector(c$redis_state$current_request,
+					    c$redis_state$current_request + 2);
+				}
+			}
+		}
+	set_state(c, T);
+
+	c$redis$cmd = command;
+	}
+
+## Gets the next response number based on a connection. This is necessary since
+## some responses may have been skipped.
+function response_num(c: connection): count
+	{
+	local resp_num = c$redis_state$current_response + 1;
+	for ( i in c$redis_state$no_response_ranges )
+		{
+		local range = c$redis_state$no_response_ranges[i];
+		assert |range| >= 1;
+		if ( |range| == 1 && resp_num > range[0] )
+			{ } # TODO: This is necessary if not using pipelining
+		if ( |range| == 2 && resp_num >= range[0] && resp_num < range[1] )
+			return range[1];
+		}
+
+	# Default: no disable/enable shenanigans
+	return resp_num;
+	}
+
+event Redis::server_data(c: connection, data: Redis::ServerData)
+	{
+	if ( ! c?$redis_state )
+		make_new_state(c);
+
+	local previous_response_num = c$redis_state$current_response;
+	c$redis_state$current_response = response_num(c);
+	set_state(c, F);
+
+	c$redis$response = data;
+	# Log each of the pending responses to this point - we will not go
+	# back.
+	while ( previous_response_num < c$redis_state$current_response )
+		{
+		if ( previous_response_num == 0 )
+			{
+			++previous_response_num;
+			next;
+			}
+
+		if ( previous_response_num in c$redis_state$pending &&
+		    c$redis_state$pending[previous_response_num]?$cmd )
+			{
+			Log::write(Redis::LOG, c$redis_state$pending[previous_response_num]);
+			delete c$redis_state$pending[previous_response_num];
+			}
+		previous_response_num += 1;
+		}
+	# Log this one if we have the request and response
+	if ( c$redis?$cmd )
+		{
+		Log::write(Redis::LOG, c$redis);
+		delete c$redis_state$pending[c$redis_state$current_response];
+		}
+	}
+
+hook finalize_redis(c: connection)
+	{
+	if ( c$redis_state$violation )
+		{
+		# If there's a violation, make sure everything gets deleted
+		delete c$redis_state;
```

Comment:
I don't believe you need this. `finalize_redis` is a removal hook, so we already know that the connection is about to get thrown away, which'll take care of the state connected to it.

Reply:
I think the point was to not log pending commands if it gets to the finalize hook, but it's kind of a roundabout way to do it. I turned it into a return